File: server/game/cards/01-Core/AClashOfKings.js
const PlotCard = require('../../plotcard.js');

class AClashOfKings extends PlotCard {
    setupCardAbilities() {
        this.reaction({
            when: {
                afterChallenge: (event) =>
                    event.challenge.winner === this.controller &&
                    event.challenge.challengeType === 'power' &&
                    event.challenge.loser.faction.power > 0
            },
            handler: () => {
                let challenge = this.game.currentChallenge;
                this.game.movePower(challenge.loser.faction, challenge.winner.faction, 1);

                this.game.addMessage(
                    "{0} uses {1} to move 1 power from {2}'s faction card to their own",
                    challenge.winner,
                    this,
                    challenge.loser
                );
            }
        });
    }
}

AClashOfKings.code = '01001';

module.exports = AClashOfKings;

File: throneteki-v2/src/Throneteki.Cards/Implementations/Packs/CoreSet/AClashOfKings.cs
using Throneteki.Cards.Abilities;
using Throneteki.Domain.Enums;
using Throneteki.Domain.Events;

namespace Throneteki.Cards.Implementations.Packs.CoreSet;

/// <summary>
/// A Clash of Kings (01001) — Plot. Income 4, Initiative 9, Claim 1, Reserve 6.
/// Reaction: After you win a Power challenge, move 1 power from the losing player's
/// faction card to your own.
/// Ported from: server/game/cards/01-Core/AClashOfKings.js
/// </summary>
[CardDefinition("01001")]
public sealed class AClashOfKings : CardScript
{
    protected override IEnumerable<CardAbilityDefinition> DeclareAbilities()
    {
        yield return AbilityBuilder.Reaction("clash-move-power")
            .Describe("Reaction: After you win a Power challenge, move 1 power from the loser's faction to yours.")
            .OnEvent<ChallengeResultDeterminedEvent>((e, state) =>
                state.ActiveChallenge?.Type == ChallengeIcon.Power && e.WinnerId != null)
            .When(ctx => CommonEffects.ControllerIsAttacker(ctx))
            .Do(ctx =>
            {
                var challenge = ctx.State.ActiveChallenge!;
                var loser = ctx.State.GetPlayer(challenge.DefendingPlayerId);
                if (loser.FactionPower <= 0) return Array.Empty<GameEvent>();

                return new GameEvent[]
                {
                    new PowerLostEvent(loser.PlayerId, PowerTargetType.Player, 1),
                    CommonEffects.GainFactionPower(ctx, 1, "A Clash of Kings"),
                };
            })
            .Build();
    }
}
